src/main/java/spoon/support/reflect/declaration/CtModuleImpl.java

```diff
@@ -0,0 +1,467 @@
+		return (T) this;
+	}
+
+	@Override
+	public <T extends CtModule> T removeRequiredModule(CtModuleRequirement requiredModule) {
+		if (requiredModule == null) {
+			return (T) this;
+		}
+		return this.removeModuleDirective(requiredModule);
+	}
+
+	@Override
+	public List<CtProvidedService> getProvidedServices() {
+		if (this.moduleDirectives.isEmpty()) {
+			return CtElementImpl.emptyList();
+		} else {
+			List<CtProvidedService> providedServices = new ArrayList<>();
+			for (CtModuleDirective moduleDirective : this.moduleDirectives) {
+				if (moduleDirective instanceof CtProvidedService) {
+					providedServices.add((CtProvidedService) moduleDirective);
+				}
+			}
+			return Collections.unmodifiableList(providedServices);
+		}
+	}
+
+	@Override
+	public <T extends CtModule> T setProvidedServices(List<CtProvidedService> providedServices) {
+		if (providedServices == null || providedServices.isEmpty()) {
+			return (T) this;
+		}
+
+		List<CtProvidedService> oldProvidedServices = getProvidedServices();
+		getFactory().getEnvironment().getModelChangeListener().onListDelete(this, CtRole.PROVIDED_SERVICE, this.moduleDirectives, new ArrayList<>(oldProvidedServices));
+		this.moduleDirectives.removeAll(oldProvidedServices);
+
+		for (CtProvidedService providedService : providedServices) {
+			this.addModuleDirective(providedService);
+		}
+
+		return (T) this;
+	}
+
+	@Override
+	public <T extends CtModule> T addProvidedService(CtProvidedService providedService) {
+		if (providedService == null) {
+			return (T) this;
+		}
+
+		this.addModuleDirective(providedService);
+		return (T) this;
+	}
+
+	@Override
+	public <T extends CtModule> T removeProvidedService(CtProvidedService providedService) {
+		if (providedService == null) {
+			return (T) this;
+		}
+		return this.removeModuleDirective(providedService);
+	}
+
+	@Override
+	public CtPackage getRootPackage() {
+		return this.rootPackage;
+	}
+
+	@Override
+	public <T extends CtModule> T setRootPackage(CtPackage rootPackage) {
+		if (rootPackage != null) {
+			rootPackage.setParent(this);
+		}
+		if (!(this instanceof ModuleFactory.CtUnnamedModule)) {
+			getFactory().getEnvironment().getModelChangeListener().onObjectUpdate(this, CtRole.SUB_PACKAGE, rootPackage, this.rootPackage);
+		}
+		this.rootPackage = rootPackage;
+		return (T) this;
+	}
+
+	@Override
+	public void accept(CtVisitor visitor) {
+		visitor.visitCtModule(this);
+	}
+
+	@Override
+	public CtModuleReference getReference() {
+		return this.getFactory().Module().createReference(this);
+	}
+
+	@Override
+	public CtModule clone() {
+		return (CtModule) super.clone();
+	}
+
+	@Override
+	public <T extends CtElement> T setParent(T parent) {
+		if (!this.getSimpleName().equals(CtModule.TOP_LEVEL_MODULE_NAME) && parent != getFactory().Module().getUnnamedModule()) {
+			throw new SpoonException("The parent of a module should necessarily be the unnamed module.");
+		}
+		return (T) super.setParent(parent);
```

Comment:
May be we can mark setParent here as Derived property and ignore call of setParent here?
I guess there is no case when somebody should be able to set different parent then `unnamedModule`...

Reply:
I agree, it will avoid unexpected behaviours there.